### PR TITLE
Support different compression types. Implement `zstd` compression support

### LINKED ISSUE #394: zstd compression support
Please add support for zstd compression.
Below is an example of compression of a csv file by zstd and gzip archivers.
For zstd, an example with the number of threads 1 and equal to the number of processors.

```
-rw-------. 1 gpadmin gpadmin 9479745682 апр 27 20:28 gpbackup.csv
-rw-rw-r--. 1 gpadmin gpadmin 1248764524 июн  5 18:43 gpzip.gz
-rw-rw-r--. 1 gpadmin gpadmin  993988596 июн  5 18:37 zstd.zst
```

```
time gzip -5 -cv gpbackup.csv > gpzip.gz
gpbackup.csv : 86.8%

real	3m5.661s
user	3m0.858s
sys	0m4.779s
```

```
time zstd -c -3 gpbackup.csv > zstd.zst
gpbackup.csv : 10.49%   (9479745682 => 993988596 bytes, /*stdout*\)

real	0m41.388s
user	0m39.149s
sys	0m5.474s
```

```
time zstd -d zstd.zst > gpbackup_0
zstd.zst            : 9479745682 bytes

real	0m24.706s
user	0m13.423s
sys	0m9.628s
```
```
time zstd -c -3 -T0 gpbackup.csv > zstd.zst
gpbackup.csv : 10.49%   (9479745682 => 993988596 bytes, /*stdout*\)

real	0m8.244s
user	0m39.601s
sys	0m4.734s
```
```
time zstd -d -T0 zstd.zst > gpbackup_0
zstd.zst            : 9479745682 bytes

real	0m20.831s
user	0m13.396s
sys     0m7.433s
```

```
lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                64
On-line CPU(s) list:   0-63
Thread(s) per core:    2
Core(s) per socket:    16
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2683 v4 @ 2.10GHz
Stepping:              1
CPU MHz:               2100.000
CPU max MHz:           2100,0000
CPU min MHz:           1200,0000
BogoMIPS:              4190.27
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              40960K
NUMA node0 CPU(s):     0-15,32-47
NUMA node1 CPU(s):     16-31,48-63
```

### PR DESCRIPTION
**Pull request by leskin-in**
1. Add a command-line option to set compression type. To disable compression, `--no-compression` or `--compression-level 0` are to be set (this is the [current](https://github.com/greenplum-db/gpbackup/tree/1a3169a615ab9137e519722a770a6c44ce7ba711) behaviour). The `--compression-type` option is optional, and `gzip` is its default value; as a result, the existing behaviour of `gpbackup` is preserved without any changes to external scripts that call it.
2. In `helper` module, `backup_helper_pipes.go` is introduced. It defines wrappers for buffer writers encapsulating various compression types (initially "no compression" and "gzip"). This allows to drop (somewhat) complex `Writer` close algorithm in `backup_helper.go`.
3. On top of the above changes, implement new compression type `zstd`. https://pkg.go.dev/github.com/klauspost/compress/zstd is used in `backup_helper` to provide this kind of compression, as well as `zstd` command-line utility in other locations.

Tests in various modules are updated accordingly. Where applicable, new test cases are added for `zstd` compression.

Closes #394 .

### CODE REVIEW

**Comment by leskin-in** [comment]
Extra (external) discussions can be found in https://github.com/arenadata/gpbackup/pull/2 and https://github.com/arenadata/gpbackup/pull/3.

**Comment by Stolb27** [comment]
These changes implement https://github.com/greenplum-db/gpbackup/issues/394

**Comment by jimmyyih** [comment]
Awesome!  This was on our TODO list for quite a while but other things kept coming up.  We'll try to review this as soon as we can but just an FYI: we're about to tag v1.21.0 very soon (any day now) so this feature would most likely go into v1.22.0 (we'll try to tag this one faster; I personally really want users to use zstd instead of gzip) and a lot of our developers will be unavailable next week.  Sorry for the inconvenience 🙏 ... but I'll try to have a full review by my EOD tomorrow.

Just curious, were any performance comparisons made between a gzip backup/restore vs. a zstd backup/restore?

**Comment by leskin-in** [comment]
@jimmyyih, thanks. Looking forward to your review.

We did not compare performance of the complete implementation between `gzip` and `zstd`.

https://github.com/greenplum-db/gpbackup/issues/394 contains comparison of performance of command-line utilities.

File: helper/backup_helper_pipes.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+// Returns errors from underlying common writer only
+func (zstdPipe ZSTDBackupPipeWriterCloser) Close() error {
+	_ = zstdPipe.zstdEncoder.Close()
+	return zstdPipe.cPipe.Close()
+}
+
+func NewZSTDBackupPipeWriterCloser(writeHandle io.WriteCloser, compressLevel int) (zstdPipe ZSTDBackupPipeWriterCloser, err error) {
+	zstdPipe.cPipe = NewCommonBackupPipeWriterCloser(writeHandle)
+	zstdPipe.zstdEncoder, err = zstd.NewWriter(zstdPipe.cPipe.bufIoWriter, zstd.WithEncoderLevel(zstd.EncoderLevel(compressLevel)))
```
Comment (jimmyyih) [inline helper/backup_helper_pipes.go:81]: `EncoderLevel` is pretty interesting from the klauspost library.  There's only 4 const enum values (`SpeedFastest`, `SpeedDefault`, `SpeedBetterCompression`, and `SpeedBestCompression`) so this line will PANIC if `compressLevel` is greater than 4.

To fix the issue, we simply need to use `zstd.EncoderLevelFromZstd(compressLevel)` instead of `zstd.EncoderLevel(compressLevel)`.

See: https://github.com/klauspost/compress/blob/v1.13.1/zstd/encoder_options.go#L180-L195

**Review by jimmyyih — COMMENTED** [review]
Finished a quick (~3 hour) review of the proposed feature.  It's very clean and thorough!  The compression refactors are good and the zstd implementation itself is good.  I ran a full CI run which came out with a green success.  Only found one small issue while going through the diffs.

I'll need to spend more time verifying if anything else needs to be changed and do my own exploratory testing of the feature (to find any weird edge cases).  As mentioned before, most of us are unavailable next week (including me) so sorry in advance for any late follow-ups.

File: helper/backup_helper_pipes.go (hunk 1 above)
Comment (leskin-in) [reply helper/backup_helper_pipes.go]: Thank you for the catch. Fixed in https://github.com/greenplum-db/gpbackup/pull/523/commits/18a33dba87ca9fc582cc19ead03a6bf9da5b3dc2.

File: integration/helper_test.go [hunk 2]
```diff
@@ -306,28 +354,58 @@ func assertErrorsHandled() {
...
+
+	if compressionType == "gzip" {
 		r, _ := gzip.NewReader(bytes.NewReader(contents))
 		contents, _ = ioutil.ReadAll(r)
-
+		Expect(string(contents)).To(Equal(expectedData))
+	} else if compressionType == "zstd" {
+		r, _ := zstd.NewReader(bytes.NewReader(contents))
+		contents, _ = ioutil.ReadAll(r)
+		Expect(string(contents)).To(Equal(expectedData))
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply integration/helper_test.go]: Refactored in https://github.com/greenplum-db/gpbackup/pull/523/commits/18a33dba87ca9fc582cc19ead03a6bf9da5b3dc2.

File: utils/util.go [hunk 3]
```diff
@@ -109,10 +109,26 @@ func ValidateFullPath(path string) error {
...
-		return errors.Errorf("Compression level must be between 1 and 9")
+func ValidateCompressionTypeAndLevel(compressionType string, compressionLevel int) error {
+	minAllowedCompressionLevels := map[string]int{
+		"gzip": 1,
+		"zstd": 1,
 	}
+	maxAllowedCompressionLevels := map[string]int{
+		"gzip": 9,
+		"zstd": 19,
+	}
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply utils/util.go]: I have also though about using a `map[string]struct`. This is introduced in https://github.com/greenplum-db/gpbackup/pull/523/commits/18a33dba87ca9fc582cc19ead03a6bf9da5b3dc2.

However, I did not introduce `Default` compression level, and would rather not do that, at least in this PR. Several problems arise:
* Currently, there is no single point where compression level is stored and set. Different pipelines for processing of this command-line option are used in `backup` and in `backup_helper`. As a result, it is not possible to introduce a single point where a default compression level could be stored;
* The parsing pipeline for `-1` compression level would be ambiguous: it would first pass the validation stage in `backup`, but then, if `backup_helper` is used, it would be run with `-1` compression level again, and would pass validation there. The reason for this is that `ValidateCompressionTypeAndLevel()` only _validates_ options, but does not _manipulate_ or _own_ their values; thus this method cannot modify the value, and we have to pass `-1` compression level around.

**Review by shivzone — APPROVED** [review]
@leskin-in nice work with addition of zstd support along with the relevant refactor.
With regards to scale/perf testing we can handle it as a separate PR (Will need to test against zstd compression here https://github.com/greenplum-db/gpbackup/blob/master/ci/scripts/scale-tests.bash and https://github.com/greenplum-db/gpbackup/blob/master/plugins/plugin_test_scale.sh)

**Review by jimmyyih — APPROVED** [review]

**Comment by jimmyyih** [comment]
Sorry for the wait.  I have manually merged into master since the Github PR "Rebase and merge" button is having some weird conflicts due to the unfortunate revert.

The master commits:
https://github.com/greenplum-db/gpbackup/commit/065fdd8f62cf5849d7c886403affe645f8ae141d
https://github.com/greenplum-db/gpbackup/commit/f2227a906d29b814595c627f36c631c7aac4f126
https://github.com/greenplum-db/gpbackup/commit/4cf71f319f7dce18855706d1f664aa8d650186e6

Thanks a lot for the contribution!
